pkg/authorization/registry/role/policybased/virtual_storage.go

```diff
@@ -0,0 +1,198 @@
+package policybased
+
+import (
+	"fmt"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	kapierrors "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/rest"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	policyregistry "github.com/openshift/origin/pkg/authorization/registry/policy"
+	roleregistry "github.com/openshift/origin/pkg/authorization/registry/role"
+)
+
+// TODO sort out resourceVersions.  Perhaps a hash of the object contents?
+
+type VirtualStorage struct {
+	policyStorage policyregistry.Registry
+}
+
+// NewVirtualStorage creates a new REST for policies.
+func NewVirtualStorage(policyStorage policyregistry.Registry) roleregistry.Storage {
+	return &VirtualStorage{policyStorage}
+}
+
+func (m *VirtualStorage) New() runtime.Object {
+	return &authorizationapi.Role{}
+}
+func (m *VirtualStorage) NewList() runtime.Object {
+	return &authorizationapi.RoleList{}
+}
+
+// TODO either add selector for fields ot eliminate the option
+func (m *VirtualStorage) List(ctx kapi.Context, label labels.Selector, field fields.Selector) (runtime.Object, error) {
+	policyList, err := m.policyStorage.ListPolicies(ctx, labels.Everything(), fields.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	roleList := &authorizationapi.RoleList{}
+
+	for _, policy := range policyList.Items {
+		for _, role := range policy.Roles {
```

Comment:
I always worry when I see a pattern that appends something (even though its not appending pointers) that uses this pattern.